x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts

```diff
@@ -20,10 +21,66 @@ import { deleteRules } from '../../rules/delete_rules';
 import { duplicateRule } from '../../rules/duplicate_rule';
 import { enableRule } from '../../rules/enable_rule';
 import { findRules } from '../../rules/find_rules';
+import { patchRules } from '../../rules/patch_rules';
+import { appplyBulkActionUpdateToRule } from '../../rules/bulk_action_update';
 import { getExportByObjectIds } from '../../rules/get_export_by_object_ids';
 import { buildSiemResponse } from '../utils';
+import { transformAlertToRule } from './utils';
+import { RuleParams } from '../../schemas/rule_schemas';
+import { FindResult } from '../../../../../../alerting/server';
 
 const BULK_ACTION_RULES_LIMIT = 10000;
+const BULK_ACTION_RULES_CHUNK = 1024;
+interface ActionPerformError {
+  error: {
+    message: string;
+    statusCode: number;
+  };
+  rule: {
+    id: string;
+    name: string;
+  };
+}
+
+type ActionPerform = undefined | ActionPerformError;
+
+// wraps bulk action and catches errors
+const actionPerformWrapper = async (
+  func: () => Promise<void>,
+  rule: FindResult<RuleParams>['data'][number]
+): Promise<ActionPerform> => {
+  try {
+    await func();
+  } catch (err) {
+    const { message, statusCode } = transformError(err);
+    return {
+      error: { message, statusCode },
+      rule: { id: rule.id, name: rule.name },
+    };
+  }
+};
+
+// to avoid large number of concurrent requests, we split request in chunks and run them sequentially
+const chunkifyRulesAction = async <T extends FindResult<RuleParams>['data'][number]>(
+  rules: T[],
+  action: (rule: T) => Promise<void>
+): Promise<ActionPerform[]> => {
+  const processChunks = async (
+    rulesChunks: T[][],
+    acc: ActionPerform[]
+  ): Promise<ActionPerform[]> => {
+    if (rulesChunks.length === 0) {
+      return acc;
+    }
+    const processed = await Promise.all(
+      rulesChunks[0].map(async (rule) => actionPerformWrapper(async () => action(rule), rule))
+    );
```

Comment:
I think using `Promise.all` is suboptimal when chunking outgoing requests. For example, let's say we're sending 50 requests in batches by 10. If one of the first ten requests takes too long to complete, it will hold back all consequent batches from being sent. So total time is 5 x (slowest request in each batch).

It could be optimized by sending a new request once one of the first ones has finished. See this PR for alternative implementation: https://github.com/elastic/kibana/pull/121110/files#diff-e34d09428e2b32ee8f5000236542eb0eee8674e6c8c1c7af9d91c2db4c5b181f